// extract-pg-schema/build/kinds/extractTable.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var R = __importStar(require("ramda"));
var vitest_1 = require("vitest");
var useSchema_1 = __importDefault(require("../tests/useSchema"));
var useTestKnex_1 = __importDefault(require("../tests/useTestKnex"));
var extractTable_1 = __importDefault(require("./extractTable"));
var makePgType = function (name, schemaName) {
    if (schemaName === void 0) { schemaName = "test"; }
    return ({
        schemaName: schemaName,
        name: name,
        kind: "table",
        comment: null,
    });
};
// const test = testWith({ schemaNames: ['test'] });
(0, vitest_1.describe)("extractTable", function () {
    var _a = (0, useTestKnex_1.default)(), getKnex = _a[0], databaseName = _a[1];
    (0, useSchema_1.default)(getKnex, "test");
    (0, vitest_1.it)("should extract simplified as well as full information_schema information", function () { return __awaiter(void 0, void 0, void 0, function () {
        var db, result, expected;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    db = getKnex();
                    return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                case 2:
                    result = _a.sent();
                    expected = {
                        name: "some_table",
                        schemaName: "test",
                        kind: "table",
                        comment: null,
                        informationSchemaValue: {
                            table_catalog: databaseName,
                            table_schema: "test",
                            table_name: "some_table",
                            table_type: "BASE TABLE",
                            self_referencing_column_name: null,
                            reference_generation: null,
                            user_defined_type_catalog: null,
                            user_defined_type_schema: null,
                            user_defined_type_name: null,
                            is_insertable_into: "YES",
                            is_typed: "NO",
                            commit_action: null,
                        },
                        indices: [],
                        checks: [],
                        columns: [
                            {
                                name: "id",
                                expandedType: "pg_catalog.int4",
                                type: { fullName: "pg_catalog.int4", kind: "base" },
                                isArray: false,
                                dimensions: 0,
                                references: [],
                                reference: null,
                                defaultValue: null,
                                indices: [],
                                isNullable: true,
                                isPrimaryKey: false,
                                generated: "NEVER",
                                isUpdatable: true,
                                isIdentity: false,
                                ordinalPosition: 1,
                                maxLength: null,
                                comment: null,
                                informationSchemaValue: {
                                    table_catalog: databaseName,
                                    table_schema: "test",
                                    table_name: "some_table",
                                    column_name: "id",
                                    ordinal_position: 1,
                                    column_default: null,
                                    is_nullable: "YES",
                                    data_type: "integer",
                                    character_maximum_length: null,
                                    character_octet_length: null,
                                    numeric_precision: 32,
                                    numeric_precision_radix: 2,
                                    numeric_scale: 0,
                                    datetime_precision: null,
                                    interval_type: null,
                                    interval_precision: null,
                                    character_set_catalog: null,
                                    character_set_schema: null,
                                    character_set_name: null,
                                    collation_catalog: null,
                                    collation_schema: null,
                                    collation_name: null,
                                    domain_catalog: null,
                                    domain_schema: null,
                                    domain_name: null,
                                    udt_catalog: databaseName,
                                    udt_schema: "pg_catalog",
                                    udt_name: "int4",
                                    scope_catalog: null,
                                    scope_schema: null,
                                    scope_name: null,
                                    maximum_cardinality: null,
                                    dtd_identifier: "1",
                                    is_self_referencing: "NO",
                                    is_identity: "NO",
                                    identity_generation: null,
                                    identity_start: null,
                                    identity_increment: null,
                                    identity_maximum: null,
                                    identity_minimum: null,
                                    identity_cycle: "NO",
                                    is_generated: "NEVER",
                                    generation_expression: null,
                                    is_updatable: "YES",
                                },
                            },
                        ],
                        isRowLevelSecurityEnabled: false,
                        isRowLevelSecurityEnforced: false,
                        securityPolicies: [],
                    };
                    (0, vitest_1.expect)(result).toStrictEqual(expected);
                    return [2 /*return*/];
            }
        });
    }); });
    (0, vitest_1.it)("should fetch column comments", function () { return __awaiter(void 0, void 0, void 0, function () {
        var db, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    db = getKnex();
                    return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, db.raw("comment on column test.some_table.id is 'id column'")];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                case 3:
                    result = _a.sent();
                    (0, vitest_1.expect)(result.columns[0].comment).toBe("id column");
                    return [2 /*return*/];
            }
        });
    }); });
    (0, vitest_1.it)("should handle arrays of primitive types", function () { return __awaiter(void 0, void 0, void 0, function () {
        var db, result, actual, expected;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    db = getKnex();
                    return [4 /*yield*/, db.raw("create table test.some_table (array_of_ints integer[], array_of_strings text[], two_dimensional_array integer[][])")];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                case 2:
                    result = _a.sent();
                    actual = R.map(R.pick(["name", "expandedType", "type", "dimensions"]), result.columns);
                    expected = [
                        {
                            name: "array_of_ints",
                            expandedType: "pg_catalog.int4[]",
                            type: { fullName: "pg_catalog.int4", kind: "base" },
                            dimensions: 1,
                        },
                        {
                            name: "array_of_strings",
                            expandedType: "pg_catalog.text[]",
                            type: { fullName: "pg_catalog.text", kind: "base" },
                            dimensions: 1,
                        },
                        {
                            name: "two_dimensional_array",
                            expandedType: "pg_catalog.int4[][]",
                            type: { fullName: "pg_catalog.int4", kind: "base" },
                            dimensions: 2,
                        },
                    ];
                    (0, vitest_1.expect)(actual).toEqual(expected);
                    return [2 /*return*/];
            }
        });
    }); });
    (0, vitest_1.it)("should fetch table checks", function () { return __awaiter(void 0, void 0, void 0, function () {
        var db, result, actual, expected;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    db = getKnex();
                    return [4 /*yield*/, db.raw("create table test.some_table_with_checks (\n        id integer constraint id_check check (id > 0),\n        products TEXT[],\n        number_of_products INT,\n        constraint products_len_check check (array_length(products, 1) = number_of_products)\n    )")];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table_with_checks"))];
                case 2:
                    result = _a.sent();
                    actual = result.checks;
                    expected = [
                        { name: "id_check", clause: "id > 0" },
                        {
                            name: "products_len_check",
                            clause: "array_length(products, 1) = number_of_products",
                        },
                    ];
                    (0, vitest_1.expect)(actual).toEqual(expected);
                    return [2 /*return*/];
            }
        });
    }); });
    (0, vitest_1.it)("should handle domains, composite types, ranges and enums as well as arrays of those", function () { return __awaiter(void 0, void 0, void 0, function () {
        var db, result, actual, expected;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    db = getKnex();
                    return [4 /*yield*/, db.raw("create domain test.some_domain as text")];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, db.raw("create type test.some_composite as (id integer, name text)")];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, db.raw("create type test.some_range as range(subtype=timestamptz)")];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, db.raw("create type test.some_enum as enum ('a', 'b', 'c')")];
                case 4:
                    _a.sent();
                    return [4 /*yield*/, db.raw("create table test.some_table (\n        d test.some_domain,\n        c test.some_composite,\n        r test.some_range,\n        e test.some_enum,\n        d_a test.some_domain[],\n        c_a test.some_composite[],\n        r_a test.some_range[],\n        e_a test.some_enum[]\n    )")];
                case 5:
                    _a.sent();
                    return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                case 6:
                    result = _a.sent();
                    actual = R.map(R.pick(["name", "expandedType", "type", "dimensions"]), result.columns);
                    expected = [
                        {
                            name: "d",
                            expandedType: "test.some_domain",
                            type: {
                                fullName: "test.some_domain",
                                kind: "domain",
                            },
                            dimensions: 0,
                        },
                        {
                            name: "c",
                            expandedType: "test.some_composite",
                            type: { fullName: "test.some_composite", kind: "composite" },
                            dimensions: 0,
                        },
                        {
                            name: "r",
                            expandedType: "test.some_range",
                            type: {
                                fullName: "test.some_range",
                                kind: "range",
                            },
                            dimensions: 0,
                        },
                        {
                            name: "e",
                            expandedType: "test.some_enum",
                            type: { fullName: "test.some_enum", kind: "enum" },
                            dimensions: 0,
                        },
                        {
                            name: "d_a",
                            expandedType: "test.some_domain[]",
                            type: {
                                fullName: "test.some_domain",
                                kind: "domain",
                            },
                            dimensions: 1,
                        },
                        {
                            name: "c_a",
                            expandedType: "test.some_composite[]",
                            type: { fullName: "test.some_composite", kind: "composite" },
                            dimensions: 1,
                        },
                        {
                            name: "r_a",
                            expandedType: "test.some_range[]",
                            type: {
                                fullName: "test.some_range",
                                kind: "range",
                            },
                            dimensions: 1,
                        },
                        {
                            name: "e_a",
                            expandedType: "test.some_enum[]",
                            type: { fullName: "test.some_enum", kind: "enum" },
                            dimensions: 1,
                        },
                    ];
                    (0, vitest_1.expect)(actual).toEqual(expected);
                    return [2 /*return*/];
            }
        });
    }); });
    (0, vitest_1.describe)("references", function () {
        (0, useSchema_1.default)(getKnex, "secondary_schema");
        (0, vitest_1.it)("should extract a simple foreign key", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create table test.linking_table (some_table_id integer references test.some_table(id))")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("linking_table"))];
                    case 3:
                        result = _a.sent();
                        expected = {
                            schemaName: "test",
                            tableName: "some_table",
                            columnName: "id",
                            onDelete: "NO ACTION",
                            onUpdate: "NO ACTION",
                            name: "linking_table_some_table_id_fkey",
                        };
                        (0, vitest_1.expect)(result.columns[0].references).toEqual([expected]);
                        // Check that deprecated version still works:
                        (0, vitest_1.expect)(result.columns[0].reference).toEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should extract a foreign key with a different schema", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table secondary_schema.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create table test.linking_table (some_table_id integer references secondary_schema.some_table(id))")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("linking_table"))];
                    case 3:
                        result = _a.sent();
                        expected = {
                            schemaName: "secondary_schema",
                            tableName: "some_table",
                            columnName: "id",
                            onDelete: "NO ACTION",
                            onUpdate: "NO ACTION",
                            name: "linking_table_some_table_id_fkey",
                        };
                        (0, vitest_1.expect)(result.columns[0].references).toEqual([expected]);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should get the onDelete and onUpdate actions", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create table test.linking_table (some_table_id integer references test.some_table(id) on delete cascade on update set null)")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("linking_table"))];
                    case 3:
                        result = _a.sent();
                        expected = {
                            schemaName: "test",
                            tableName: "some_table",
                            columnName: "id",
                            onDelete: "CASCADE",
                            onUpdate: "SET NULL",
                            name: "linking_table_some_table_id_fkey",
                        };
                        (0, vitest_1.expect)(result.columns[0].references).toEqual([expected]);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    (0, vitest_1.describe)("indices", function () {
        (0, vitest_1.it)("should extract a simple index", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create index some_table_id_idx on test.some_table (id)")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        expected = [
                            {
                                name: "some_table_id_idx",
                                isPrimary: false,
                                isUnique: false,
                                columns: [
                                    {
                                        name: "id",
                                        definition: "id",
                                    },
                                ],
                            },
                        ];
                        (0, vitest_1.expect)(result.indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should extract a unique index", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create unique index some_table_id_idx on test.some_table (id)")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        expected = [
                            {
                                name: "some_table_id_idx",
                                isPrimary: false,
                                isUnique: true,
                                columns: [
                                    {
                                        name: "id",
                                        definition: "id",
                                    },
                                ],
                            },
                        ];
                        (0, vitest_1.expect)(result.indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("it should extract a primary key", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 2:
                        result = _a.sent();
                        expected = [
                            {
                                name: "some_table_pkey",
                                isPrimary: true,
                                isUnique: true,
                                columns: [
                                    {
                                        name: "id",
                                        definition: "id",
                                    },
                                ],
                            },
                        ];
                        (0, vitest_1.expect)(result.indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should extract a multi-column index", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer, kind integer)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create index some_table_id_idx on test.some_table (id, kind)")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        expected = [
                            {
                                name: "some_table_id_idx",
                                isPrimary: false,
                                isUnique: false,
                                columns: [
                                    {
                                        name: "id",
                                        definition: "id",
                                    },
                                    {
                                        name: "kind",
                                        definition: "kind",
                                    },
                                ],
                            },
                        ];
                        (0, vitest_1.expect)(result.indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should extract a functional index", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create index some_table_id_idx on test.some_table (abs(id))")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        expected = [
                            {
                                name: "some_table_id_idx",
                                isPrimary: false,
                                isUnique: false,
                                columns: [
                                    {
                                        name: null,
                                        definition: "abs(id)",
                                    },
                                ],
                            },
                        ];
                        (0, vitest_1.expect)(result.indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    (0, vitest_1.describe)("row-level security", function () {
        (0, vitest_1.it)("should extract isRowLevelSecurityEnabled", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("alter table test.some_table enable row level security")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        (0, vitest_1.expect)(result.isRowLevelSecurityEnabled).toEqual(true);
                        (0, vitest_1.expect)(result.isRowLevelSecurityEnforced).toEqual(false);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should extract isRowLevelSecurityEnforced", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("alter table test.some_table force row level security")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        (0, vitest_1.expect)(result.isRowLevelSecurityEnabled).toEqual(false);
                        (0, vitest_1.expect)(result.isRowLevelSecurityEnforced).toEqual(true);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    (0, vitest_1.describe)("securityPolicies", function () {
        (0, vitest_1.it)("should extract empty array when no policy is defined", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 2:
                        result = _a.sent();
                        expected = [];
                        (0, vitest_1.expect)(result.securityPolicies).toEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("it should extract a simple security policy", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create policy test_policy on test.some_table")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 3:
                        result = _a.sent();
                        expected = [
                            {
                                name: "test_policy",
                                isPermissive: true,
                                rolesAppliedTo: ["public"],
                                commandType: "ALL",
                                visibilityExpression: null,
                                modifiabilityExpression: null,
                            },
                        ];
                        (0, vitest_1.expect)(result.securityPolicies).toEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("it should extract a complex security policy", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create role test_role")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create policy test_policy on test.some_table as restrictive for update to test_role, postgres using (id = 1) with check (true)")];
                    case 3:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 4:
                        result = _a.sent();
                        expected = [
                            {
                                name: "test_policy",
                                isPermissive: false,
                                rolesAppliedTo: ["postgres", "test_role"],
                                commandType: "UPDATE",
                                visibilityExpression: "(id = 1)",
                                modifiabilityExpression: "true",
                            },
                        ];
                        (0, vitest_1.expect)(result.securityPolicies).toEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
    });
    (0, vitest_1.describe)("bugfixes", function () {
        (0, vitest_1.it)("should not report duplicate columns when a column has multiple foreign key constraints", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer primary key)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create table test.linking_table (\n          some_table_id integer,\n          constraint \"fk_1\" foreign key (\"some_table_id\") references test.some_table(id),\n          constraint \"fk_2\" foreign key (\"some_table_id\") references test.some_table(id)\n        )")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("linking_table"))];
                    case 3:
                        result = _a.sent();
                        (0, vitest_1.expect)(result.columns).toHaveLength(1);
                        (0, vitest_1.expect)(result.columns[0].references).toStrictEqual([
                            {
                                schemaName: "test",
                                tableName: "some_table",
                                columnName: "id",
                                onDelete: "NO ACTION",
                                onUpdate: "NO ACTION",
                                name: "fk_1",
                            },
                            {
                                schemaName: "test",
                                tableName: "some_table",
                                columnName: "id",
                                onDelete: "NO ACTION",
                                onUpdate: "NO ACTION",
                                name: "fk_2",
                            },
                        ]);
                        // Check deprecated version still works:
                        (0, vitest_1.expect)(result.columns[0].reference).toEqual({
                            schemaName: "test",
                            tableName: "some_table",
                            columnName: "id",
                            onDelete: "NO ACTION",
                            onUpdate: "NO ACTION",
                            name: "fk_1",
                        });
                        return [2 /*return*/];
                }
            });
        }); });
        (0, vitest_1.it)("should not extract indices from another schema", function () { return __awaiter(void 0, void 0, void 0, function () {
            var db, result, expected;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        db = getKnex();
                        return [4 /*yield*/, db.raw("create table test.some_table (id integer)")];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create index some_table_id_idx on test.some_table (id)")];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create schema test2")];
                    case 3:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create table test2.some_table (id integer)")];
                    case 4:
                        _a.sent();
                        return [4 /*yield*/, db.raw("create index some_table_id_idx2 on test2.some_table (id)")];
                    case 5:
                        _a.sent();
                        return [4 /*yield*/, (0, extractTable_1.default)(db, makePgType("some_table"))];
                    case 6:
                        result = _a.sent();
                        expected = [
                            {
                                isPrimary: false,
                                name: "some_table_id_idx",
                            },
                        ];
                        (0, vitest_1.expect)(result.columns[0].indices).toStrictEqual(expected);
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
